### PR TITLE
DAOS-15145 pool: add pool collective function

### PR DESCRIPTION
**Pull request by wangdi1**
Add pool collective function to skip down and downout targets.

Required-githooks: true
Features: rebuild

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/container/srv_container.c [hunk 1]
```diff
@@ -1817,8 +1817,9 @@ ds_cont_tgt_refresh_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid,
 	uuid_copy(arg.cont_uuid, cont_uuid);
 	arg.min_eph = eph;
 
-	rc = dss_task_collective(cont_refresh_vos_agg_eph_one, &arg,
-				 DSS_ULT_FL_PERIODIC);
+	rc = ds_pool_task_collective(pool_uuid, PO_COMP_ST_NEW | PO_COMP_ST_DOWN |
+				     PO_COMP_ST_DOWNOUT, cont_refresh_vos_agg_eph_one,
+				     &arg, DSS_ULT_FL_PERIODIC);
```
Comment (liw) [inline src/container/srv_container.c:1822]: I think we should add a macro to src/include/daos_srv/pool.h:

```
/* Complement of the map states of targets that make up a pool shard */
#define DS_POOL_TGT_MAP_STATES_EX (PO_COMP_ST_NEW | PO_COMP_ST_DOWN | PO_COMP_ST_DOWNOUT)
```

And use the macro in most cases. I can do this in one of my PRs if you agree.

**Review by liw — DISMISSED** [review]

File: src/container/srv_container.c (hunk 1 above)
Comment (wangdi1) [reply src/container/srv_container.c:1822]: @liw  Sure, please fix this in your PR. Thanks!

**Review by liuxuezhao — DISMISSED** [review]

**Comment by wangdi1** [comment]
ping

**Review by NiuYawei — CHANGES_REQUESTED** [review]
Could you add the "Features: DaosCoreTestNvme" in commit message?

**Review by liw — DISMISSED** [review]

**Comment by wangdi1** [comment]
> Could you add the "Features: DaosCoreTestNvme" in commit message?

sure

**Review by NiuYawei — DISMISSED** [review]

**Review by NiuYawei — DISMISSED** [review]

**Review by liw — DISMISSED** [review]

File: src/pool/srv_target.c [hunk 2]
```diff
@@ -2205,17 +2193,100 @@ pool_child_discard(void *data)
...
+				goto out;
+		}
+	}
+
+	if (thread)
+		rc = dss_thread_collective_reduce(coll_ops, coll_args, flags);
+	else
+		rc = dss_task_collective_reduce(coll_ops, coll_args, flags);
+
+	DL_CDEBUG(rc == 0, DB_MD, DLOG_ERR, rc, DF_UUID " collective", DP_UUID(pool_uuid));
```
Comment (NiuYawei) [inline src/pool/srv_target.c:2227]: This error message caused NLT failure.

Comment (wangdi1) [reply src/pool/srv_target.c]: sure. thanks.

**Review by NiuYawei — DISMISSED** [review]

**Review by liw — DISMISSED** [review]

**Review by liuxuezhao — DISMISSED** [review]

**Comment by wangdi1** [comment]
sigh, hardware large did not get to run due to some hardware issue.

**Review by NiuYawei — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Pull request merged**